public/app/features/alerting/unified/rule-list/PaginatedDataSourceLoader.tsx

```diff
@@ -0,0 +1,128 @@
+import { useEffect, useMemo, useRef } from 'react';
+
+import { Dropdown, Icon, IconButton, Menu, Stack, Text } from '@grafana/ui';
+import { DataSourceRuleGroupIdentifier, ExternalRulesSourceIdentifier, RuleGroup } from 'app/types/unified-alerting';
+
+import { hashRule } from '../utils/rule-id';
+
+import { AlertRuleLoader } from './AlertRuleLoader';
+import { DataSourceSection, DataSourceSectionProps } from './components/DataSourceSection';
+import { LazyPagination } from './components/LazyPagination';
+import { ListGroup } from './components/ListGroup';
+import { ListSection } from './components/ListSection';
+import { usePrometheusGroupsGenerator } from './hooks/prometheusGroupsGenerator';
+import { usePaginatedPrometheusGroups } from './hooks/usePaginatedPrometheusGroups';
+
+export const GROUP_PAGE_SIZE = 40;
+
+interface PaginatedDataSourceLoaderProps extends Required<Pick<DataSourceSectionProps, 'application'>> {
+  rulesSourceIdentifier: ExternalRulesSourceIdentifier;
+}
+export function PaginatedDataSourceLoader({ rulesSourceIdentifier, application }: PaginatedDataSourceLoaderProps) {
+  const { uid, name } = rulesSourceIdentifier;
+  const prometheusGroupsGenerator = usePrometheusGroupsGenerator();
+
+  const groupsGenerator = useRef(prometheusGroupsGenerator(rulesSourceIdentifier, GROUP_PAGE_SIZE));
+
+  useEffect(() => {
+    const currentGenerator = groupsGenerator.current;
+    return () => {
+      currentGenerator.return();
+    };
+  }, [groupsGenerator]);
+
+  const {
+    page: groupsPage,
+    nextPage,
+    previousPage,
+    canMoveForward,
+    canMoveBackward,
+    isLoading,
+  } = usePaginatedPrometheusGroups(groupsGenerator.current, GROUP_PAGE_SIZE);
+
+  const groupsByNamespace = useMemo(() => groupBy(groupsPage, 'file'), [groupsPage]);
+
+  return (
+    <DataSourceSection name={name} application={application} uid={uid} isLoading={isLoading}>
+      <Stack direction="column" gap={1}>
+        {Object.entries(groupsByNamespace).map(([namespace, groups]) => (
+          <ListSection
+            key={namespace}
+            title={
+              <Stack direction="row" gap={1} alignItems="center">
+                <Icon name="folder" />{' '}
+                <Text variant="body" element="h3">
+                  {namespace}
+                </Text>
+              </Stack>
+            }
+          >
+            {groups.map((group) => (
+              <RuleGroupListItem
+                key={`${String(rulesSourceIdentifier.uid)}-${namespace}-${group.name}`}
+                group={group}
+                rulesSourceIdentifier={rulesSourceIdentifier}
+                namespaceName={namespace}
+              />
+            ))}
+          </ListSection>
+        ))}
+        <LazyPagination
+          nextPage={nextPage}
+          previousPage={previousPage}
+          canMoveForward={canMoveForward}
+          canMoveBackward={canMoveBackward}
+        />
+      </Stack>
+    </DataSourceSection>
```

Comment:
same here, do you think we could create a reusable component for data source and grafana managed loaders?

Reply:
I'm not sure. If you take a closer look. you'll notice that the only thing we could actually reuse is the `<Lazy pagination` block
The `ListSection` construction varies between DS and Grafana managed rules. The same for `RuleGroupListItem`
For GMA we have `GrafanaRuleGroupListItem` which takes a bit different props and internally is quite different